File: src/http/admin/tracks.go
package admin

import (
	"net/http"
	"tapesonic/model"
	"tapesonic/search"
	"tapesonic/util"

	"github.com/google/uuid"
)

type TrackRsArtist struct {
	Id   uuid.UUID
	Name string
}

type TrackRs struct {
	Id string

	SourceId *string
	RemoteId *string

	Artist *TrackRsArtist
	Title  string

	ArtworkId *uuid.UUID
}

func libraryToTrackRs(track model.LibraryTrack) TrackRs {
	trackRs := TrackRs{
		Id:        track.Id,
		Title:     track.Title,
		ArtworkId: track.ArtworkId,
	}

	if track.SourceId != nil {
		sourceId := track.SourceId.String()
		trackRs.SourceId = &sourceId
	}
	if track.RemoteId != nil {
		remoteId := track.RemoteId.String()
		trackRs.RemoteId = &remoteId
	}

	if track.ArtistId != nil {
		trackRs.Artist = &TrackRsArtist{
			Id:   *track.ArtistId,
			Name: track.ArtistName,
		}
	}

	return trackRs
}

func GetTracks(auth *authenticator, search *search.SearchService) WebappHandler {
	return func(r *http.Request) (any, error) {
		user, err := auth.Authenticate(r)
		if err != nil {
			return nil, err
		}

		query := r.URL.Query().Get("q")

		tracks, err := search.FindTracksByQuery(user.Id, query)
		if err != nil {
			return nil, err
		}

		return util.Map(tracks, libraryToTrackRs), nil
	}
}

File: webapp/src/api/tracks.ts
import api from '@/api/api';

export interface TrackRsArtist {
    Id: string;
    Name: string;
}

export interface TrackRs {
    Id: string;

    SourceId: string | null;
    RemoteId: string | null;

    Artist: TrackRsArtist | null;
    Title: string;

    ThumbnailId: string | null;
}

export default {
    async searchTracks(query: string): Promise<TrackRs[]> {
        const params = new URLSearchParams({ "q": query });
        const response = await fetch(`/api/tracks?${params}`, { method: "GET" });
        return await api.handleRequiredResponse(response);
    },
}
